core/src/main/scala/org/apache/spark/util/SparkConfRegisterLoader.scala

```diff
@@ -0,0 +1,61 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.util
+
+import java.net.URL
+
+import scala.collection.JavaConverters._
+import scala.io.Source
+import scala.reflect.runtime.{universe => ru}
+import scala.util.control.NonFatal
+
+import org.apache.spark.internal.Logging
+
+object SparkConfRegisterLoader extends Logging {
```

Comment:
Why do we need this class for initing configs eagerly? We cannot simply load them as `PythonSQLUtils` does?
https://github.com/apache/spark/blob/d99d0d27be875bba692bcfe376f90c930e170380/sql/core/src/main/scala/org/apache/spark/sql/api/python/PythonSQLUtils.scala#L48-L58

Reply:
@maropu, awesome! I was not aware of this in `PythonSQLUtils`. But I'm still worried about the maintenance if other external modules also have registered SQLConf. And this code assumes the hive module is already included which is not suitable for all other moudles.

Reply:
Yea, we might be able to brush up it based on the current implementation. But, IMO it'd be better to avoid loading classes from a resource file.

Reply:
Yea it's easier to just list all the known config objects, but I agree with @linhongliu-db that some modules are optional and we should only load the objects from modules when the modules are loaded. FYI all modules under the `external` directory are not loaded by default, and if they start to define configs, we will need this framework. It's also true that other spark vendors may define their own configs.

This is similar to the data source registry resource files, so I don't have a big concern here as this is not the first time.

Reply:
We should let `PythonSQLUtils` use this new framework though.

Reply:
This looks an overkill. So do we plan to separate the options for `external` modules? I would just simply do `ScalaReflection.mirror.staticModule("org.apache.spark.sql.hive.HiveUtils")` in `SetCommand`.

Reply:
This allows people to define SQL configs outside of Spark, e.g. spark packages, vendor-specific configs. We should probably make it clear in the PR title and description. If we just want to fix the bug, we can hardcode the loading of `HiveUtis` class like @HyukjinKwon mentioned.

How about we separate this PR into 2:
1. One PR to add `SparkConfRegisterLoader`, which can be used in SetCommand and `PythonSQLUtils`. This fixes the bug for SetCommand
2. One PR to add the resource file loading, which allows people to define SQL configs outside of Spark. We can discuss more if this feature is reasonable in that PR.

Reply:
@cloud-fan sure, I can do this

Reply:
> One PR to add the resource file loading, which allows people to define SQL configs outside of Spark.

We should probably make `SQLConf.buildConf` and `SQLConf.buildStaticConf` as proper developer APIs first

Reply:
> One PR to add the resource file loading, which allows people to define SQL configs outside of Spark.

Ah, I see. Anyway, I also think It is better to separate the bug fix and the feature improvement as @cloud-fan said above.

Reply:
Fixing `SetCommand` first sounds good to me.

Reply:
thanks all for reviewing, I'll update this PR to focus on bug fixing only. and open another to discuss resource file loading.